src/geometry.cpp

```diff
@@ -536,6 +536,204 @@ extern "C" double openmc_calculate_optical_thickness(
+    p.event_calculate_xs();
+
+    total_sigma_t += p.macro_xs().total;
+    total_sigma_a += p.macro_xs().absorption;
+    valid_samples++;
+  }
+
+  if (valid_samples > 0) {
+    *sigma_t_output = total_sigma_t / valid_samples;
+    *sigma_a_output = total_sigma_a / valid_samples;
+  } else {
+    *sigma_t_output = 0.0;
+    *sigma_a_output = 0.0;
+  }
+}
+
+extern "C" void openmc_get_inter_voxel_transport_operator(
+  Position start_voxel_min, Position start_voxel_max, Position end_voxel_min,
+  Position end_voxel_max, int num_rays, double* transport_operator_output)
+{
+  double total_transport = 0.0;
+  SpatialBox start_box(start_voxel_min, start_voxel_max);
+  SpatialBox end_box(end_voxel_min, end_voxel_max);
+  int64_t id = 1;
+  uint64_t seed = init_seed(id, STREAM_SOURCE);
+  int valid_samples = 0;
+
+#pragma omp parallel for reduction(+ : total_transport, valid_samples)
+
+  for (int i = 0; i < num_rays; ++i) {
+    Position start_sampled_position = start_box.sample(&seed);
+    Position end_sampled_position = end_box.sample(&seed);
+
+    double optical_thickness = openmc_calculate_optical_thickness(
+      start_sampled_position, end_sampled_position);
+
+    Position ray = end_sampled_position - start_sampled_position;
+    double length_squared = ray.dot(ray);
+
+    double transport = std::exp(-optical_thickness) / length_squared;
+    total_transport += transport;
+    valid_samples++;
+  }
+
+  if (valid_samples > 0) {
+    *transport_operator_output = total_transport / valid_samples;
+  } else {
+    *transport_operator_output = 0.0;
+  }
+}
+
+Direction sample_unit_sphere(uint64_t* seed)
+{
+  double phi = 2.0 * PI * prn(seed);
+  double cos_theta = 2.0 * prn(seed) - 1.0;
+  double sin_theta = std::sqrt(1.0 - cos_theta * cos_theta);
+
+  Direction dir;
+  dir.x = sin_theta * std::cos(phi);
+  dir.y = sin_theta * std::sin(phi);
+  dir.z = cos_theta;
+
+  return dir;
+}
+
+double distance_to_exit(const Position& point, const Direction& direction,
+  const Position& voxel_min, const Position& voxel_max)
+{
+  // compute the distance to exit the voxel by checking the intersection with
+  // each of the six faces of the voxel and minimizing the positive value
+
+  double t_min = std::numeric_limits<double>::infinity();
+
+  if (std::abs(direction.x) > FP_PRECISION) {
+    double neg_x = (voxel_min.x - point.x) / direction.x;
+    double pos_x = (voxel_max.x - point.x) / direction.x;
+
+    t_min = std::min(t_min, (direction.x > 0) ? pos_x : neg_x);
+  }
+
+  if (std::abs(direction.y) > FP_PRECISION) {
+    double neg_y = (voxel_min.y - point.y) / direction.y;
+    double pos_y = (voxel_max.y - point.y) / direction.y;
+
+    t_min = std::min(t_min, (direction.y > 0) ? pos_y : neg_y);
+  }
+
+  if (std::abs(direction.z) > FP_PRECISION) {
+    double neg_z = (voxel_min.z - point.z) / direction.z;
+    double pos_z = (voxel_max.z - point.z) / direction.z;
+
+    t_min = std::min(t_min, (direction.z > 0) ? pos_z : neg_z);
+  }
```

Comment:
[performance: minor?]
```suggestion
  double bound = 0;
  if (std::abs(direction.x) > FP_PRECISION) {
    bound = (direction.x > 0) ? voxel_max.x : voxel_min.x;
    t_min = std::min(t_min, (bound - point.x) / direction.x);
  }

  if (std::abs(direction.y) > FP_PRECISION) {
    bound = (direction.y > 0) ? voxel_max.y : voxel_min.y;
    t_min = std::min(t_min, (bound - point.y) / direction.y);
  }

  if (std::abs(direction.z) > FP_PRECISION) {
    bound = (direction.z > 0) ? voxel_max.z : voxel_min.z;
    t_min = std::min(t_min, (bound - point.z) / direction.z);
  }
```

Reply:
Good point. No need to store negative distance.